File: src/electron/sessions.ts
import crypto from 'node:crypto';

export interface SessionProfile {
	id: number;
	name: string;
	login: string;
	role: string;
}

export interface ActiveSession {
	token: string;
	profile: SessionProfile;
	issuedAt: number;
}

export class SessionStore {
	private sessions = new Map<string, ActiveSession>();

	create(profile: SessionProfile): ActiveSession {
		const token = crypto.randomUUID();
		const session: ActiveSession = {
			token,
			profile,
			issuedAt: Date.now(),
		};
		this.sessions.set(token, session);
		return session;
	}

	get(token: string | undefined | null): ActiveSession | undefined {
		if (!token) return undefined;
		return this.sessions.get(token);
	}

	require(token: string | undefined | null): ActiveSession {
		const session = this.get(token);
		if (!session) {
			throw new Error('Sessão inválida. Faça login novamente.');
		}
		return session;
	}

	revoke(token: string): void {
		this.sessions.delete(token);
	}

	revokeAll(): void {
		this.sessions.clear();
	}
}

File: src-tauri/src/sessions.rs
use std::collections::HashMap;

use chrono::Utc;
use std::sync::RwLock;
use uuid::Uuid;

use crate::types::UserProfile;

#[derive(Clone)]
pub struct ActiveSession {
    pub token: String,
    pub profile: UserProfile,
    #[allow(dead_code)]
    pub(crate) issued_at: i64,
}

#[derive(Default)]
pub struct SessionStore {
    sessions: RwLock<HashMap<String, ActiveSession>>,
}

impl SessionStore {
    pub fn create(&self, profile: UserProfile) -> ActiveSession {
        let token = Uuid::new_v4().to_string();
        let session = ActiveSession {
            token: token.clone(),
            profile,
            issued_at: Utc::now().timestamp_millis(),
        };
        // Unwrap is safe here as we are not handling lock poisoning in this simple app
        self.sessions
            .write()
            .unwrap()
            .insert(token.clone(), session.clone());
        session
    }

    pub fn get(&self, token: &str) -> Option<ActiveSession> {
        self.sessions.read().unwrap().get(token).cloned()
    }

    pub fn require(&self, token: &str) -> Result<ActiveSession, &'static str> {
        self.get(token)
            .ok_or("Sessão inválida. Faça login novamente.")
    }

    pub fn revoke(&self, token: &str) {
        self.sessions.write().unwrap().remove(token);
    }
}
